src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -40,346 +40,528 @@ Shard::Shard(
-            {
-                legacy_ = true;
-                return fail("incompatible, missing backend final key");
-            }
+            // Ensure all threads get the same object
+            pCache_->canonicalize_replace_client(hash, nodeObject);
+            fetchReport.wasFound = true;
 
-            // Check final key's value
-            SerialIter sIt(nObj->getData().data(), nObj->getData().size());
-            if (sIt.get32() != version)
-                return fail("invalid version");
+            // Since this was a 'hard' fetch, we will log it
+            JLOG(j_.trace()) << "HOS: " << hash << " fetch: in shard db";
+        }
+    }
 
-            if (sIt.get32() != firstSeq_ || sIt.get32() != lastSeq_)
-                return fail("out of range ledger sequences");
+    return nodeObject;
+}
 
-            if (sIt.get256().isZero())
-                return fail("invalid last ledger hash");
+bool
+Shard::fetchNodeObjectFromCache(
+    uint256 const& hash,
+    std::shared_ptr<NodeObject>& nodeObject)
+{
+    auto const backendCount{makeBackendCount()};
+    if (!backendCount)
+        return false;
 
-            if (exists(dir_ / LgrDBName) && exists(dir_ / TxDBName))
-                final_ = true;
+    nodeObject = pCache_->fetch(hash);
+    if (nodeObject || nCache_->touch_if_exists(hash))
+        return true;
+    return false;
+}
 
-            backendComplete_ = true;
-        }
+bool
+Shard::storeLedger(
+    std::shared_ptr<Ledger const> const& srcLedger,
+    std::shared_ptr<Ledger const> const& next)
+{
+    if (state_ != acquire)
+    {
+        // Ignore residual calls from InboundLedgers
+        JLOG(j_.trace()) << "shard " << index_ << " not acquiring";
+        return false;
     }
-    catch (std::exception const& e)
+    if (containsLedger(srcLedger->info().seq))
     {
-        return fail(
-            std::string("exception ") + e.what() + " in function " + __func__);
+        JLOG(j_.debug()) << "shard " << index_ << " ledger already stored";
+        return false;
     }
 
-    setBackendCache(lock);
-    if (!initSQLite(lock))
-        return fail({});
+    if (srcLedger->info().hash.isZero() ||
+        srcLedger->info().accountHash.isZero())
+    {
+        JLOG(j_.error()) << "source ledger seq " << srcLedger->info().seq
+                         << " is invalid";
+        return false;
+    }
+    auto& srcDB{const_cast<Database&>(srcLedger->stateMap().family().db())};
+    if (&srcDB == &(app_.getShardFamily()->db()))
+    {
+        JLOG(j_.error()) << "source and destination databases are the same";
+        return false;
+    }
 
-    setFileStats(lock);
-    return true;
-}
+    auto const backendCount{makeBackendCount()};
+    if (!backendCount)
+        return false;
 
-boost::optional<std::uint32_t>
-Shard::prepare()
-{
-    std::lock_guard lock(mutex_);
-    assert(backend_);
+    Batch batch;
+    batch.reserve(batchWritePreallocationSize);
+    size_t storedSize{0};
```

Comment:
It looks like `storedSize` is unused?  Consider removing it from here and stop tracking it in `storeBatch`, below.

Reply:
This should be reported to the caller of `storeledger`, will fix!